libs/wire-api-federation/src/Wire/API/Federation/Version.hs

```diff
@@ -66,6 +81,107 @@ instance ToSchema VersionInfo where
+        element "until" VersionRangeTagUntil,
+        element "from-until" VersionRangeTagFromUntil
+      ]
+
+versionPairSchema :: ValueSchema NamedSwaggerDoc (Version, Version)
+versionPairSchema =
+  object "VersionPair" $
+    (,)
+      <$> fst .= field "from" schema
+      <*> snd .= field "until" schema
+
+data VersionRange
+  = AllVersions
+  | FromVersion Version
+  | UntilVersion Version
+  | FromUntilVersion Version Version
+
+deriving instance Eq VersionRange
+
+deriving instance Show VersionRange
+
+deriving instance Ord VersionRange
+
+makePrisms ''VersionRange
+
+instance ToSchema VersionRange where
+  schema =
+    object "VersionRange" $
+      fromTagged
+        <$> toTagged
+          .= bind
+            (fst .= field "tag" versionRangeTagSchema)
+            (snd .= fieldOver _1 "value" untaggedSchema)
+    where
+      toTagged :: VersionRange -> (VersionRangeTag, VersionRange)
+      toTagged d@AllVersions = (VersionRangeTagAll, d)
+      toTagged d@(FromVersion _) = (VersionRangeTagFrom, d)
+      toTagged d@(UntilVersion _) = (VersionRangeTagUntil, d)
+      toTagged d@(FromUntilVersion _ _) = (VersionRangeTagFromUntil, d)
+
+      fromTagged :: (VersionRangeTag, VersionRange) -> VersionRange
+      fromTagged = snd
+
+      untaggedSchema = dispatch $ \case
+        VersionRangeTagAll -> tag _AllVersions null_
+        VersionRangeTagFrom -> tag _FromVersion (unnamed schema)
+        VersionRangeTagUntil -> tag _UntilVersion (unnamed schema)
+        VersionRangeTagFromUntil -> tag _FromUntilVersion $ unnamed versionPairSchema
+
+deriving via Schema VersionRange instance ToJSON VersionRange
+
+deriving via Schema VersionRange instance FromJSON VersionRange
+
+-- | Compute the lower and upper boundary of a version range. The first
+-- component of the pair is the lower boundary, while the second component is
+-- the upper boundary.
+versionRangeToBoundaries :: VersionRange -> (Version, Version)
+versionRangeToBoundaries AllVersions = (minBound @Version, maxBound @Version)
+versionRangeToBoundaries (FromVersion fv) = (fv, maxBound @Version)
+versionRangeToBoundaries (UntilVersion uv) = (minBound @Version, pred uv)
+versionRangeToBoundaries (FromUntilVersion fv uv) = (fv, pred uv)
+
+-- | Checks if a version is within a given version range.
+isWithinRange :: Version -> VersionRange -> Bool
+isWithinRange v vr =
+  let (lo, hi) = versionRangeToBoundaries vr
+   in lo <= v && v <= hi
+
+-- | For a version range of a local backend and for a set of versions that a
+-- remote backend supports, compute the newest version supported by both. The
+-- remote versions are given as integers as the range of versions supported by
+-- the remote backend can include a version unknown to the local backend. If
+-- there is no version in common, the return value is 'Nothing'.
+versionInCommon :: VersionRange -> Set Int -> Maybe Version
+versionInCommon localVersions remoteVersions =
+  foldl' f Nothing (Set.map inRange remoteVersions)
```

Comment:
You could also compute a `Set Int` from localVersions and then take the maximum of the intersection of the two sets, but I don't know if that's easier.